File: course/section_3_sliding_window/code/04-permutation-in-string.py
from collections import Counter


def pr[T](val: T, debug=True) -> T:
    if debug:
        print(val)
    return val


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        pattern = Counter(s1)
        k = len(s1)
        current_window = Counter(s2[:k])
        if pattern == current_window:
            return True
        for i in range(k, len(s2)):
            entering = s2[i]
            leaving = s2[i - k]
            current_window[entering] = current_window.get(entering, 0) + 1
            current_window[leaving] = current_window.get(leaving) - 1  # type: ignore
            if current_window[leaving] == 0:
                del current_window[leaving]
            if pattern == current_window:
                return True

        return False

File: course/section_3_sliding_window/code/04-permutation-in-string.php
<?php 

class Solution {

    function checkInclusion($s1, $s2) {

        $n = strlen($s1);
        $m = strlen($s2);

        if ($n > $m) {
            return false;
        }

        $pattern = [];
        $window = [];

        for ($i = 0; $i < $n; $i++) {
            $pattern[$s1[$i]] = ($pattern[$s1[$i]] ?? 0) + 1;
        }

        for ($i = 0; $i < $n; $i++) {
            $window[$s2[$i]] = ($window[$s2[$i]] ?? 0) + 1;
        }

        if ($window == $pattern) {
            return true;
        }

        for ($i = $n; $i < $m; $i++) {

            $entering = $s2[$i];
            $leaving = $s2[$i - $n];

            $window[$entering] = ($window[$entering] ?? 0) + 1;
            $window[$leaving]--;

            if ($window[$leaving] <= 0) {
                unset($window[$leaving]);
            }

            if ($window == $pattern) {
                return true;
            }
        }

        return false;
    }
}